LSM: Fixes: Grid.release_at_checkpoint; Tree.TableMutable.ValuesCache

sentientwaffle opened the pull request:
These two changes are mostly unrelated, except they both hinge on the `SetAssociativeCache`.

## `Grid.release_at_checkpoint()`

Fix crash on `if (verify) assert(self.search(set, key) == null);`.

Explanation of bug:

1. (During compaction)
2. The first level B table is fully buffered, so release its blocks (`Grid.release()`).
    1. Release the block addresses from the free set at the next checkpoint (not yet).
    2. Remove the blocks from the `SetAssociativeCache` (right now).
3. (Keep compacting the rest of level B)
4. While compaction is still running, the `StateMachine` handles a request that queries the block we just released. This adds it back to the cache.
5. After checkpoint, the released block is reassigned. We try to write the new version to the cache (`put_no_clobber`), but there is already an (old) entry there!
(We want to release the blocks mid-compaction rather than later because the index block is in memory — we don't want to risk missing the cache later and having to do IO to get it back.)

Explanation of fix:

1. In `Grid.release()`, instead of removing the entry from the `Grid` cache (we may still need it after all), just knock its "score" in the `SetAssociativeCache` down.
2. Relax the `SetAssociativeCache.search` assertion.

## `Tree.TableMutable.ValuesCache`

- Change the values cache from a `HashMap` to a `SetAssociativeCache`.
- Hook the cache up so that it is updated and correctly accessed.
- The cache is only for reads from the current snapshot.
- Move the values cache into the `TableMutable` because they are accessed similarly:
  - both are synchronous (in-memory)
  - both are only checked when reading from the current snapshot
  - both must not have out-of-date values
- Change `prefetch_clear()` to `prefetch_setup(snapshot)`. We were already calling it at the beginning of a commit rather than the end to simplify the `StateMachine` unit test structure.
  - Right now the state machine always passes `snapshot_latest`.
- Refactor `Groove.lookup` and prefetching to keep the in-memory table access (cache, mutable, immutable) all in one place. The caller is responsible for checking the in-memory lookup before falling back to the (on-disk) levels.
- Fix the `values_cache` comment, which previously (incorrectly) indicated that the value cache was shared between multiple types.

## Don't read unwritten tables

This fixes an issue where tables' blocks were being read from disk before the block was ever written.
This is what happens:
1. Compaction is running...
2. Compaction marks input tables with `snapshot_max = compaction_snapshot`.
3. Compaction marks output tables with `snapshot_min = compaction_snapshot` (and `snapshot_max = intMax(u64)`).
4. But the tables' snapshots are updated before the compaction finishes! And before the blocks are written to disk. This is important because it means compaction doesn't need to buffer the changes.
  - Aside: This isn't a new bug, but it was hidden until the "ManifestLevel: SegmentedArray binary search" (https://github.com/tigerbeetledb/tigerbeetle/pull/165) refactor removed the need for batched manifest updates.

Fix:
- Add `Tree.compacted_snapshot_max()`:
  - Don't ever query a higher snapshot — this function handles the logic of "Which snapshot is securely on disk and safe to query?"
  - Only query the mutable table when `snapshot == tree.compacted_snapshot_max()`.
  - When a tree recovers from a checkpoint, it needs to know what op was last committed (`superblock.working.vsr_state.commit_min`). `Tree.init()` now requires that the superblock is opened. The tests are updated accordingly.
- `snapshot_min` is now *exclusive*.
- `snapshot_max` is now *inclusive*.
- Don't ever `compact(op:0)` (even in the `lsm/test.zig`). `replica.zig` doesn't. (This allows a tighter assertion in `compact()`).
- `prefetch_setup`'s op is now optional, to allow the tree to use the current snapshot (which isn't known by the state machine).
- Fix a typo in `PostedGroove`.

## Don't remove invisible tables until new tables are ready

Summary:
*Do not remove (compaction input) invisible tables until prefetch() can query the new (compaction output) tables.*

Details:
- Track `compaction_op_done` explicitly, rather than deriving it from `compaction_op`. Not only is this approach simpler and easier to follow, it grants the tree more control over when prefetch can switch to using the newer snapshot.
- Avoid duplicate `compact(op)` calls. Compacting the same op twice due to a crash can lead to nondeterministic storage when combined with crash/recover.
  - Document the relationship between `compaction_op`, `compaction_op_done`, and their behavior after recovery from checkpoint.
- Don't remove invisible tables until all sub-compactions of a half-measure compaction have finished. In more detail:
  - Previously each full compaction shared a snapshot; now each half-compaction has its own snapshot (see `compaction_snapshot_for_op`). This allows the output tables of the first half to be used for prefetch() during the second half. My reasoning is that we want to start using the new tables as soon as it is safe to do so, since they are warm in the cache. If we wait until the end of the full compaction, they may have been evicted by old tables.
  - Previously we removed the invisible tables as each level's compaction finished, rather than removing all invisible tables together when all of the levels' compactions finish. But we essentially need to atomically switch from the old tables to the new, by updating the prefetch snapshot from the old `compaction_op_done` to the new `compaction_op`.

Misc fixes:
- Rename `Compaction.remove_level_a` to `Compaction.input_level_a`. Compaction will not always result in the old table's removal. And even when it does, that removal is not immediate (and not the `Compaction`'s responsibility). I want to disambiguate between "remove a table" and "make a table invisible by updating its snapshot_max".
- Update some outdated comments.

sentientwaffle committed:
LSM: Grid.release: Don't remove block from cache

Fix crash on `if (verify) assert(self.search(set, key) == null);`.

Explanation of bug:

1. (During compaction)
2. The first level B table is fully buffered, so release its blocks (`Grid.release()`).
    1. Release the block addresses from the free set at the next checkpoint (not yet).
    2. Remove the blocks from the `SetAssociativeCache` (right now).
3. (Keep compacting the rest of level B)
4. While compaction is still running, the `StateMachine` handles a request that queries the block we just released. This adds it back to the cache.
5. After checkpoint, the released block is reassigned. We try to write the new version to the cache (`put_no_clobber`), but there is already an (old) entry there!
(We want to release the blocks mid-compaction rather than later because the index block is in memory — we don't want to risk missing the cache later and having to do IO to get it back.)

Explanation of fix:

1. In `Grid.release()`, instead of removing the entry from the `Grid` cache (we may still need it after all), just knock its "score" in the `SetAssociativeCache` down.
2. Relax the `SetAssociativeCache.search` assertion.

sentientwaffle committed:
LSM: Tree ValuesCache

- Change the values cache from a `HashMap` to a `SetAssociativeCache`.
- Hook the cache up so that it is updated and correctly accessed.
- The cache is only for reads from the current snapshot.
- Move the values cache into the `TableMutable` because they are accessed similarly:
  - both are synchronous (in-memory)
  - both are only checked when reading from the current snapshot
  - both must not have out-of-date values
- Change `prefetch_clear()` to `prefetch_setup(snapshot)`. We were already calling it at the beginning of a commit rather than the end to simplify the `StateMachine` unit test structure.
  - Right now the state machine always passes `snapshot_latest`.
- Refactor `Groove.lookup` and prefetching to keep the in-memory table access (cache, mutable, immutable) all in one place. The caller is responsible for checking the in-memory lookup before falling back to the (on-disk) levels.
- Fix the `values_cache` comment, which previously (incorrectly) indicated that the value cache was shared between multiple types.

sentientwaffle committed:
LSM: TableMutable: Avoid redundant values in cache

sentientwaffle committed:
LSM: Don't read unwritten tables

This fixes an issue where tables' blocks were being read from disk before the block was ever written.
This is what happens:
1. Compaction is running...
2. Compaction marks input tables with `snapshot_max = compaction_snapshot`.
3. Compaction marks output tables with `snapshot_min = compaction_snapshot` (and `snapshot_max = intMax(u64)`).
4. But the tables' snapshots are updated before the compaction finishes! And before the blocks are written to disk. This is important because it means compaction doesn't need to buffer the changes.
  - Aside: This isn't a new bug, but it was hidden until the "ManifestLevel: SegmentedArray binary search" (https://github.com/tigerbeetledb/tigerbeetle/pull/165) refactor removed the need for batched manifest updates.

Fix:
- Add `Tree.compacted_snapshot_max()`:
  - Don't ever query a higher snapshot — this function handles the logic of "Which snapshot is securely on disk and safe to query?"
  - Only query the mutable table when `snapshot == tree.compacted_snapshot_max()`.
  - When a tree recovers from a checkpoint, it needs to know what op was last committed (`superblock.working.vsr_state.commit_min`). `Tree.init()` now requires that the superblock is opened. The tests are updated accordingly.
- `snapshot_min` is now *exclusive*.
- `snapshot_max` is now *inclusive*.
- Don't ever `compact(op:0)` (even in the `lsm/test.zig`). `replica.zig` doesn't. (This allows a tighter assertion in `compact()`).
- `prefetch_setup`'s op is now optional, to allow the tree to use the current snapshot (which isn't known by the state machine).
- Fix a typo in `PostedGroove`.

Lastly, as part of this fix, I wrote up quite a bit of documentation around snapshots, but it is in the `dj-lsm-review` branch with the rest of my LSM documentation updates to avoid merge conflicts in `compaction.zig`.

sentientwaffle committed:
LSM: Remove invisible tables after compaction done

Summary:
*Do not remove (compaction input) invisible tables until prefetch() can query the new (compaction output) tables.*

Details:
- Track `compaction_op_done` explicitly, rather than deriving it from `compaction_op`. Not only is this approach simpler and easier to follow, it grants the tree more control over when prefetch can switch to using the newer snapshot.
- Avoid duplicate `compact(op)` calls. Compacting the same op twice due to a crash can lead to nondeterministic storage when combined with crash/recover.
  - Document the relationship between `compaction_op`, `compaction_op_done`, and their behavior after recovery from checkpoint.
- Don't remove invisible tables until all sub-compactions of a half-measure compaction have finished. In more detail:
  - Previously each full compaction shared a snapshot; now each half-compaction has its own snapshot (see `compaction_snapshot_for_op`). This allows the output tables of the first half to be used for prefetch() during the second half. My reasoning is that we want to start using the new tables as soon as it is safe to do so, since they are warm in the cache. If we wait until the end of the full compaction, they may have been evicted by old tables.
  - Previously we removed the invisible tables as each level's compaction finished, rather than removing all invisible tables together when all of the levels' compactions finish. But we essentially need to atomically switch from the old tables to the new, by updating the prefetch snapshot from the old `compaction_op_done` to the new `compaction_op`.

Misc fixes:
- Rename `Compaction.remove_level_a` to `Compaction.input_level_a`. Compaction will not always result in the old table's removal. And even when it does, that removal is not immediate (and not the `Compaction`'s responsibility). I want to disambiguate between "remove a table" and "make a table invisible by updating its snapshot_max".
- Update some outdated comments.

jorangreef commented on src/config.zig:
```diff
@@ -48,22 +48,22 @@ pub const memory_size_max_default = 1024 * 1024 * 1024;
 /// The maximum number of two-phase transfers to store in memory:
 /// This impacts the amount of memory allocated at initialization by the server.
-pub const transfers_pending_max = transfers_max;
+pub const transfers_pending_max = std.math.floorPowerOfTwo(usize, transfers_max);
```
Let's rather bring these in directly as powers of 2, e.g. `67_108_864` or `64 * 1024 * 1024`, without flooring, so that one can see the magnitude more clearly, and so that we don't need to take a std lib dependency.

jorangreef replied:
We could also entirely drop the switch on `deployment_environment` now if you like, since we'll be moving to runtime config for these soon?

sentientwaffle replied:
I will leave it for now, and remove it when the runtime config is set up.

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -110,7 +110,7 @@ pub fn CompactionType(
         manifest: *Manifest,
         level_b: u8,
-        remove_level_a: ?TableInfo,
+        input_level_a: ?TableInfo,
```
```suggestion
        level_a_table: ?TableInfo,
```

jorangreef replied:
Sometimes comments give good clues as to naming things, e.g. later on we refer to this as "level_a table":

`// Mark the level_a table as invisible if it was provided;`

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -450,15 +454,15 @@ pub fn CompactionType(
+            // Mark the level_a table as invisible if it was provided;
+            // it has been merged into level_b.
             // TODO: Release the grid blocks associated with level_a as well
-            if (compaction.level_b != 0) {
```
We could still bring this in as an assertion if the level_a table exists, so that we bind the value of the `level_a_table` optional with respect to the value of `level_b`.

This would then be the opposite of the `level_b == 0` assertion in the else branch.

jorangreef replied:
On second thought, I guess we do have this assertion here implicitly, in the form of checked arithmetic, since we subtract 1 from `level_b`.

sentientwaffle replied:
Yes, this change was just a small refactor to remove some redundant code.

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -156,17 +156,20 @@ pub fn GridType(comptime Storage: type) type {
             grid.assert_not_reading(address, null);
 
-            grid.cache.remove(address); // Free the way in the cache now to reduce conflict misses.
+            grid.cache.demote(address);
```
Great name.

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -337,6 +340,17 @@ pub fn GridType(comptime Storage: type) type {
+                // 2. All of the Grid's Read IOPS are occupied, so queue the read.
+                // 3. The replica checkpoints.
+                // 4. The read dequeues, but the requested block is no longer allocated.
+                // TODO(State Transfer) How does a non-repairable block trigger state transfer?
```
Regarding triggering state transfer, I've been thinking:

1. If a local read results in a fault, then the replica should attempt a remote read.
2. If a remote replica has the block then it responds (and the local read completes), otherwise it nacks.
3. If we receive too many nacks or if we get the feeling that we are too far behind (perhaps the primary nacks), then complete the read callback but now with a null result, so that it unwinds the stack all the way back to VSR, which then initiates state transfer. At present, we expect that reads always return a block, so to support this bubbling up, we'll need to make the block result optional.

sentientwaffle replied:
I'm glad that is how it will work! I actually started to make this exact change before, but backed it out because I figured there was a different plan that matched the existing signature :grin:. I will update this TODO, but defer making the "block → optional-block" callback refactor until I implement state transfer.

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -426,12 +430,11 @@ pub fn GrooveType(
 
-            objects_cache.* = .{};
-            try objects_cache.ensureTotalCapacity(allocator, cache_size);
+            objects_cache.* = try ObjectTree.TableMutable.ValuesCache.init(allocator, cache_size);
```
Should we rename `cache_size` here also to `cache_cardinality_max` or to `cache_entries_max`?

sentientwaffle replied:
Yes, once I rebase this branch onto `main` this will be `cache_entries_max`.

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -481,7 +483,7 @@ pub fn GrooveType(
                     node_pool,
                     grid,
-                    null, // No value cache for index trees.
+                    null, // No values cache for index trees.
```
```suggestion
                    null, // No values cache for index trees, since they only do range queries.
```

To explain why.

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -541,35 +544,49 @@ pub fn GrooveType(
+                groove.prefetch_objects.clearRetainingCapacity();
+            } else {
+                // If there is a leftover snapshot, then prefetch() was never called, so there
+                // must already be no queued objects or ids.
```
Can we assert this?

jorangreef replied:
Can we clarify what "leftover snapshot" means? I'm not sure what this is?

sentientwaffle replied:
I will clarify the comment. But we actually do assert this just a couple lines down — the assertions cover both branches of this condition. The difference is that in the first branch we have to clear the `prefetch_objects`, but in the second branch `prefetch_objects` is already empty):
```
            if (groove.prefetch_snapshot == null) {
                groove.prefetch_objects.clearRetainingCapacity();
            } else {
                // If there is a leftover snapshot, then prefetch() was never called, so there
                // must already be no queued objects or ids.
            }

            groove.prefetch_snapshot = snapshot_target;
            assert(groove.prefetch_objects.count() == 0);
            assert(groove.prefetch_ids.count() == 0);
```

The comment just explains why we don't need to clear `prefetch_objects` when a `prefetch_snapshot` is already set.

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -541,35 +544,49 @@ pub fn GrooveType(
+            // We may query the input tables of an ongoing compaction, but must not query the
+            // output tables until the compaction is complete. (Until then, the output tables may
+            // be in the manifest but not yet on disk).
+            const snapshot_max = groove.ids.compaction_op_done;
```
Great comment above here.

Is `compaction_op_done` the snapshot that will be used for the output tables?

jorangreef replied:
Could we set `snapshot_max` to `compaction_op_done - 1` here so that `snapshot_min/max` values are always inclusive on both bounds?

Making bounds inclusive for the WAL helped with the repair logic there at the time, and it would be great if we can always think of bounds across VSR/LSM the same way.

sentientwaffle replied:
> Is `compaction_op_done` the snapshot that will be used for the output tables?

No, output tables are assigned `compaction_snapshot_for_op(compaction_op)` as their snapshot. `compaction_op_done` is set to `compaction_op` when the compaction finishes.

But per your suggestion about inclusive snapshot min/max, I am going to rework this a bit.

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -541,35 +544,49 @@ pub fn GrooveType(
+            if (groove.ids.lookup_from_memory(groove.prefetch_snapshot.?, id)) |id_tree_value| {
+                if (id_tree_value.tombstone()) {
+                    // Do nothing; a prefetched ID not present in prefetch_objects indicates
+                    // that the object has either been deleted or never existed.
```
```suggestion
                    // Do nothing; an explicit ID tombstone indicates that the object was deleted.
```

jorangreef replied:
There should be no tombstone if the object never existed.

And if the object never existed and was nevertheless explicitly deleted by inserting a tombstone, then the new comment is still accurate?

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -541,35 +544,49 @@ pub fn GrooveType(
+                    // Do nothing; a prefetched ID not present in prefetch_objects indicates
+                    // that the object has either been deleted or never existed.
+                } else {
+                    const object = groove.objects.table_mutable.get(id_tree_value.timestamp).?;
```
I think we also need to use `lookup_from_memory()` here since the object may be in the immutable table.

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -648,18 +668,18 @@ pub fn GrooveType(
                 // If not in the LSM tree's cache, the object must be read from disk and added
                 // to the auxiliary prefetch_objects hash map.
                 // TODO: this LSM tree function needlessly checks the LSM tree's cache a
                 // second time. Adding API to the LSM tree to avoid this may be worthwhile.
```
```suggestion
                // TODO This LSM tree function needlessly checks the LSM tree's cache a
                // second time. Consider adding an API to the LSM tree to avoid this.
```

sentientwaffle replied:
This TODO can be removed; `lookup_from_levels` is exactly the LSM tree API that this "TODO" describes :tada:.

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -648,18 +668,18 @@ pub fn GrooveType(
                 // TODO: this LSM tree function needlessly checks the LSM tree's cache a
                 // second time. Adding API to the LSM tree to avoid this may be worthwhile.
-                worker.context.groove.ids.lookup(
+                worker.context.groove.ids.lookup_from_levels(
```
```suggestion
                worker.context.groove.ids.lookup_from_disk(
```

Corresponds to `lookup_from_memory`?

jorangreef replied:
`lookup_from_levels` is also good though.

sentientwaffle replied:
I named it `lookup_from_levels()` because it doesn't necessarily hit the disk at all — all of the necessary blocks may already be in the grid's cache. I'd rather find a better name for `lookup_from_memory`, but haven't come up with one yet.

jorangreef commented on src/lsm/manifest.zig:
```diff
@@ -50,21 +50,27 @@ pub fn TableInfoType(comptime Table: type) type {
             //
-            // For example, Compaction may update the snapshot_max of tables during removal to make 
+            // For example, Compaction may update the snapshot_max of tables during removal to make
             // them invisible. It will also scan for tables that are invisible with the same snapshot.
```
We don't need to reference how snapshots were previously done.

Would be great to improve the detail of this comment overall some more.

jorangreef commented on src/lsm/manifest.zig:
```diff
@@ -50,21 +50,27 @@ pub fn TableInfoType(comptime Table: type) type {
-            // - inclusive to snapshot_min (new tables are inserted with snapshot=snapshot_min)
-            // - exclusive to snapshot_max (tables are removed / made invisible by setting snapshot_max)
-            
+            // - exclusive to snapshot_min (new tables are inserted with snapshot=snapshot_min)
```
Can we keep this as inclusive, and then rather adjust the compaction snapshot that is assigned to `snapshot_min` by one?

jorangreef replied:
It follows also what we do in Journal with bounds there, if a table is visible to `snapshot_min..snapshot_max` and both bounds are inclusive.

No need to remember which is the inclusive bound and which is the exclusive bound.

jorangreef commented on src/lsm/posted_groove.zig:
```diff
@@ -90,7 +91,10 @@ pub fn PostedGrooveType(comptime Storage: type) type {
         prefetch_objects: PrefetchObjects,
 
-        /// This field is necessary to expose the same open()/compact_cpu()/compact_io() function
+        /// The snapshot to prefetch from.
```
Is the optional type used as a guard only, or is it valid also not to specify a snapshot?

sentientwaffle replied:
It is valid to pass `null` to `prefetch_setup`, but `prefetch_setup` converts a `null` to "the current snapshot" (chosen based on the state of the ongoing compaction). So the `prefetch_snapshot` property is only optional as a guard.

jorangreef commented on src/lsm/posted_groove.zig:
```diff
@@ -176,10 +180,24 @@ pub fn PostedGrooveType(comptime Storage: type) type {
+                // must already be no queued objects or ids.
+            }
+
+            groove.prefetch_snapshot = snapshot_target;
```
How can we share all this logic across the Grooves?

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -176,6 +177,16 @@ pub fn SetAssociativeCache(
+            const way = self.search(set, key) orelse return;
+
+            const count = self.counts.get(set.offset + way);
+            self.counts.set(set.offset + way, div_ceil(count, 2));
```
If I understand correctly, this will set a count of 3 to 2, 2 to 1, 1 to 1, without setting the count to 0 (that would remove the key from the set)?

What about just setting the count to 1 always when demote is called?

jorangreef replied:
Doing so would reduce the probability that another element is evicted, when this one should be next to go.

In fact, we could even tweak the clock hand to point here.

jorangreef replied:
I see `demote()` less as decrement the hit rate, and more like, mark this as highly replaceable.

sentientwaffle replied:
I'll set the count to 1, but I'm going to leave the clock hand alone for predictability. (And I think it could actually be worse when a batch of blocks are demoted if we move the clock hand to the last one.)

jorangreef replied:
Spot on about not messing with the clock hand.

jorangreef commented on src/lsm/table_mutable.zig:
```diff
@@ -20,11 +21,47 @@ pub fn TableMutableType(comptime Table: type) type {
+        /// The values cache is updated (in bulk) when the mutable table is sorted and frozen,
+        /// rather than updating on every `put()`/`remove()`.
+        /// This avoids the redundancy of storing duplicate values between the mutable table
+        /// and the values cache.
```
```suggestion
        /// This amortizes cache inserts for hot keys in the mutable table, and
        /// avoids redundantly storing duplicate values in both the mutable table and values cache.
```

jorangreef commented on src/lsm/table_mutable.zig:
```diff
@@ -20,11 +21,47 @@ pub fn TableMutableType(comptime Table: type) type {
         dirty: bool = false,
 
-        pub fn init(allocator: mem.Allocator, commit_count_max: u32) !TableMutable {
+        /// This is used to accelerate point lookups and is not used for range queries.
+        /// Secondary index trees used only for range queries can therefore set this to null.
+        ///
+        /// The values cache is only used for the latest snapshot for simplicity.
+        /// Earlier snapshots will still be able to utilize the block cache.
+        ///
+        /// The values cache is updated (in bulk) when the mutable table is sorted and frozen,
+        /// rather than updating on every `put()`/`remove()`.
```
Really nice to read this documentation.

jorangreef commented on src/lsm/table_mutable.zig:
```diff
@@ -110,6 +155,12 @@ pub fn TableMutableType(comptime Table: type) type {
+                for (values) |*value| {
+                    cache.insert(key_from_value(value)).* = value.*;
+                }
+            }
```
For performance, it seems like we should iterate the values only once, copying them into `values_max` and into `values_cache` in the same branch.

For example, if the mutable table is 64 MiB, then we're moving values between L1/L2/L3 cache twice instead of once.

jorangreef replied:
This means we need to unwrap `table.values_cache` on each iteration, but that branch should be easily predicted, since it depends only on the tree and not the workload.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -1,3 +1,68 @@
+//!   ┼───┬───┼───┬───┼
+//!   0 2 4 6 8 0 2 4 6
+//!
+//! Notice how in the checkpoint recovery example above, we are careful not to `compact(op)` twice
+//! for any op (even if we crash/recover), since that could lead to differences between replicas'
+//! storage. The last measure of `commit()`s is always only in memory, so it is safe to repeat.
```
This is an awesome diagram DJ.

While it is possible to carefully skip compactions before the checkpoint, and while we should document this, I don't think that we should:

* The mental model with our ratcheting is supposed to be that we simply replay work from the commit log after a checkpoint, and that doing so is deterministic, that the entire async hook lifecycle is processed for every op, to rebuild the same in-memory state. It's the nature of a log and a state snapshot.
* Whereas if we skip some of the compacts for ops in the commit log after a crash, this masks any underlying non-deterministic bugs should they exist. In fact, I think that we may actually have a bug in how we acquire from the free set, where these acquires are persisted too early during a checkpoint to which they don't relate.
* It also increases the probability that our in-memory state might get out of sync, for example if we crash exactly at a checkpoint, because skipping some compactions means we need to set all the resulting state correctly, instead of just deriving it via replay which is simpler.
* It introduces bi-modality in performance, also potentially in callback synchronicity.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -112,6 +167,19 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
+        /// * is equal to `compaction_op` while no compaction is in progress
+        ///   (at init; between half-measures).
+        /// * is less than `compaction_op` while any compaction is in progress.
+        compaction_op_done: u64,
```
When seeing this in our prefetch code, I was wondering if we couldn't improve the name, to be more along the lines of "the highest snapshot that is safe to prefetch from", e.g. `prefetch_snapshot_max`.

sentientwaffle replied:
:grin: I used that exact name in an earlier iteration! I changed it to `compaction_op_done` to emphasize the relationship with `compaction_op`. But I'll most likely change it back to `prefetch_snapshot_max` (or similar) as part of the inclusive snapshot_min refactor.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -149,16 +219,24 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
+            // The commits from the measure following commit_min were in the mutable table, and
+            // thus not preserved in the checkpoint.
+            // But the corresponding `compact()` updates were preserved, and must not be repeated
+            // to ensure determinstic storage.
```
If everything is deterministic as it should be, then I believe it should be safe to replay `compact()` calls?

jorangreef replied:
And then we should do that, in order to expose any non-determinism.

sentientwaffle replied:
Per https://github.com/tigerbeetledb/tigerbeetle/pull/166#discussion_r979409817, it is safe to replay `compact()` calls only if the results from the previous `compact()` calls were not preserved. So we either need to not replay `compact()` calls for the last measure before a checkpoint trigger (which is what I originally implemented), or not preserve the compactions of the last measure before a checkpoint trigger within the compaction (which is what you suggest in the referenced comment).

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -193,33 +262,42 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
+        ) void {
+            assert(snapshot <= tree.compaction_op_done);
+            if (config.verify) {
+                // The caller is responsible for checking the mutable table.
+                assert(tree.lookup_from_memory(snapshot, key) == null);
```
Nice!

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -701,28 +787,76 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
-                assert(tree.compaction_table_immutable.status == .idle);
-            } else {
+            const finished_even_levels = compaction_beat == half_measure_beat_count - 1;
+            const finished_odd_levels = compaction_beat == config.lsm_batch_multiple - 1;
```
```suggestion
            const compacted_levels_even = compaction_beat == half_measure_beat_count - 1;
            const compacted_levels_odd = compaction_beat == config.lsm_batch_multiple - 1;
```

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -701,28 +787,76 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
         }
 
+        /// Called at the end of each compaction tick.
         fn compact_done(tree: *Tree) void {
```
@kprotty Would you please review everything here in `compact_done()` since you are most familiar?

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -899,6 +1016,26 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
     };
 }
 
+/// Returns the target snapshot of a compaction tick at the given op.
+///
+/// This snapshot will be:
+/// - the `snapshot_min` of a compaction's output tables, and
+/// - the `snapshot_max` of a compaction's input tables.
+///
+/// The target snapshot of a compaction is the last op of the preceding half-measure.
+///
+/// At the start of the current batch, mutable table inserts from the previous batch
+/// would be in the immutable table. This means the current batch compaction will
+/// actually be flushing to disk (levels) mutable table updates from the previous batch.
+///
+/// -1 as the ops are zero based so the "last" op from previous batch is reflected.
+pub fn compaction_snapshot_for_op(op: u64) u64 {
+    // Each half-measure has a snapshot instead of each full measure, because this allows the
+    // output tables of the first half-measure's compaction to be prefetched against earlier —
+    // hopefully while they are still warm in the cache from being written.
+    return std.mem.alignBackward(op, @divExact(config.lsm_batch_multiple, 2)) -| 1;
+}
+
```
Awesome docs and great to encapsulate this in one place.

jorangreef reviewed (commented):
Awesome work @sentientwaffle.

kprotty commented on src/lsm/tree.zig:
```diff
@@ -701,28 +787,76 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
+                const callback = tree.compaction_callback.?;
+                tree.compaction_callback = null;
+                callback(tree);
+                return;
```
Could this code be moved into a `compact_finish()` function which is also called by `manifest_checkpoint_callback()`?

sentientwaffle committed:
LSM: Code review fixes

sentientwaffle committed:
LSM: Make snapshot_min inclusive

- Make `snapshot_min` inclusive (to match `snapshot_max`).
  - This is simpler to reason about, visualize, and explain.
- Document+diagram compaction's affect on (and interaction with) snapshots.
- Diagram `op_checkpoint` vs `op_checkpoint_trigger` and the interaction with WAL wraps, since it is easier to understand visually.
- Fix a few comment typos.
- Rename `compaction_op_done` to `prefetch_snapshot_max`. Both are accurate names, but the latter matches its usage more often.

sentientwaffle committed:
LSM: Manifest: Add more checks

kprotty commented on src/lsm/tree.zig:
```diff
@@ -727,28 +813,74 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type, comptime tree_
-            var still_compacting = false;
+            // At the end of every beat:
+            // - ensure mutable table can be flushed to immutable table.
+            defer assert(tree.table_mutable.can_commit_batch(tree.options.commit_entries_max));
```
Ran the LSM test with `accounts_to_insert_per_op` equal to `8191` or `forest_options.accounts.tree_options_id.commit_entries_max` (using `tree_options_object`'s max would overcommit the IdTree's mutable table).

On `op=7`, this assert is hit in the IdTree with its mutable table count at `32764` while the commit_entries_max is `8191`. I assume this is due to running after `compact_finish()` or `compact_drive()` from `defer`. Moving this assertion into `compact_finish()` before the callback is invoked fixes it.

sentientwaffle replied:
Great catch! I'll move the assertion to `compact_finish()` as you suggested.

sentientwaffle committed:
LSM: Tree: Move assert to compact_finish

The assert must be run at the end of a beat. But previously it would also run in the `still_compacting` branch.

sentientwaffle committed:
LSM: Queue partial blocks for compaction on open()

See invariant: "Compaction must compact partially full blocks".

sentientwaffle committed:
VSR: Skip client table updates after recover

When recovering from a checkpoint, don't update the client table until we are past the previous `op_checkpoint_trigger` (since those updates are already in the client table). We can only verify this when `if (self.client_table().get(prepare.header.client))`, since the entry may have been evicted. (In which case we still skip it, to preserve the determinism).

sentientwaffle committed:
LSM: Groove: Fix coordination between caches

A value may be in the id cache, but not the object cache (or vice-versa).

sentientwaffle committed:
SuperBlock: Fix client table body decoding

`bodies.len` is zero when every reply has `header.size == @sizeOf(vsr.Header)`.
For example, if all of the requests were `register`s.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2669,10 +2669,28 @@ pub fn ReplicaType(
-            if (reply.header.operation == .register) {
-                self.create_client_table_entry(reply);
+            if (self.op_checkpoint == 0 or
+                prepare.header.op > self.op_checkpoint + config.lsm_batch_multiple)
```
It would be good to encapsulate this condition.

jorangreef replied:
So that we can then use it also for deciding whether to `compact()`?

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2669,10 +2669,28 @@ pub fn ReplicaType(
+                if (self.client_table().get(prepare.header.client)) |entry| {
+                    assert(entry.reply.header.command == .reply);
+                    assert(entry.reply.header.op >= prepare.header.op);
+                }
```
Is there anything else we can check here?

For example, if the client table does not have an entry for this client, then we assume it must have been evicted, but can we then assert that the client table is full?

jorangreef commented on src/lsm/manifest_log.zig:
```diff
@@ -248,6 +248,10 @@ pub fn ManifestLogType(comptime Storage: type, comptime TableInfo: type) type {
                 }
             }
 
+            if (block_entry_count(block) < entry_count_max) {
+                manifest.queue_for_compaction(block_reference.address);
+            }
```
Nice catch!

jorangreef commented on src/lsm/manifest_log.zig:
```diff
@@ -434,6 +434,9 @@ pub fn ManifestLogType(comptime Storage: type, comptime TableInfo: type) type {
 
+            // Compact a single manifest block — to minimize latency spikes, we want to do the bare
+            // minimum of compaction work required.
+            // TODO Compact more than 1 block if fragmentation is outstripping the compaction rate.
```
Perfect explanation of fragmentation vs compaction rates.

jorangreef commented on src/lsm/manifest.zig:
```diff
@@ -191,7 +203,9 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
 
-            // TODO Verify that tables can be found exactly before returning.
+            if (config.verify) {
+                assert(manifest_level.contains(table));
```
Awesome! Reads so well.

jorangreef commented on src/lsm/manifest_log.zig:
```diff
@@ -437,7 +439,6 @@ pub fn ManifestLogType(comptime Storage: type, comptime TableInfo: type) type {
                 assert(block.address > 0);
 
                 manifest_log.reading = true;
-                manifest_log.read_callback = callback;
```
Much better—I'm surprised I did it this way, setting to `undefined` only to set the callback again in the first branch.

jorangreef commented on src/lsm/manifest.zig:
```diff
@@ -49,20 +49,20 @@ pub fn TableInfoType(comptime Table: type) type {
         pub fn visible(table: *const TableInfo, snapshot: u64) bool {
             assert(table.address != 0);
-            assert(table.snapshot_min < table.snapshot_max);
+            assert(table.snapshot_min <= table.snapshot_max);
```
This means that a table may be compacted in the very next half-measure after it was created?

sentientwaffle replied:
Yes, I think that is possible. If a compaction is from level N → N+1, then the next half-measure's compaction is from level N+1 → N+2.

jorangreef reviewed (dismissed):
Awesome work @sentientwaffle !

sentientwaffle committed:
LSM: Refactor op_compacted()

jorangreef commented on src/vsr/superblock.zig:
```diff
@@ -140,6 +140,15 @@ pub const SuperBlockSector = extern struct {
+        /// But the corresponding `compact()` updates were preserved, and must not be repeated
+        /// to ensure determinstic storage.
+        pub fn op_compacted(state: VSRState, op: u64) bool {
+            return state.commit_min > 0 and op <= state.commit_min + config.lsm_batch_multiple;
```
What happens if we checkpoint at the earliest possible place in the WAL that we could checkpoint (`0 + batch_multiple`), so that `commit_min` is exactly 0?
